core/dbt/parser/manifest.py

```diff
@@ -220,12 +220,23 @@ def load(self):
         # of parsers to lists of file strings. The file strings are
         # used to get the SourceFiles from the manifest files.
         start_read_files = time.perf_counter()
-        project_parser_files = {}
         saved_files = {}
         if self.saved_manifest:
             saved_files = self.saved_manifest.files
-        for project in self.all_projects.values():
-            read_files(project, self.manifest.files, project_parser_files, saved_files)
+
+        project_parser_files = {
+            project.project_name: read_files(project, saved_files)
+            for project in self.all_projects.values()
+        }
+
+        for files in project_parser_files.values():
+            for file in files:
+                saved_files[file.file_id] = file
+        project_parser_files = {
```

Comment:
I am not _really_ happy with this dict comprehension. 

My recommendation would be to refactor `project_parser_files` to have `list[SourceFile]` as values instead of `list[str]` (which are the file ids). It would make the code below simpler as that is that is fetching the `SourceFile` from the `saved_files` all the time through the `file_id`. Memory wise it should not matter _too much_, since the `saved_files` and `project_parser_files` will have _pointers_ to the same `SourceFile`.

Reply:
while we are at it: a refactor to use `Project` instead of `str` (`Project.project_name`) as keys might also make sense